docs/strings.md

```diff
@@ -0,0 +1,464 @@
+The core string class should implement the minimum required for generic implementation of string operations, which can be implemented either using dynamic dispatch over the string interface or as generic reifications over the concrete types.
+We should make it possible to support a rich Unicode library that uses the same string objects as the standard library (and third-party libraries that use the same interface) and provides functions for iteration with unicode graphemes, locale-aware collation and other more complex operations that not all programs need.
+
+The interface for strings would look roughly like this:
+
+```verona
+/**
+ * String contains only 7-bit ASCII characters and is stored with each code
+ * point in an 8-bit integer.
+ */
+class StringEncodingASCII
+{
+	typedef CharType = U8;
+}
+/**
+ * String is stored as UTF-8-encoded text.
+ */
+class StringEncodingUTF8
+{
+	typedef CharType = U8;
+}
+/**
+ * String is stored as UTF-16-encoded text.
+ */
+class StringEncodingUTF16
+{
+	typedef CharType = U16;
+}
+/**
+ * String is stored as UTF-32-encoded text.
+ */
+class StringEncodingUTF32
+{
+	typedef CharType = U32;
+}
+/**
+ * Interface for other string encodings.  Any other type can be used as a valid
+ * string encoding but code that does not have any special cases for that
+ * encoding will be able to treat is as unknown.
+ */
+interface StringEncodingCustom
+{
+	typedef CharType;
+}
+
+/**
+ * The set of Unicode encodings.  All standard library functions that have
+ * special cases for specific encodings should handle these.
+ */
+typedef UnicodeStringEncoding = StringEncodingASCII |
+                                StringEncodingUTF8 |
+                                StringEncodingUTF16 |
+                                StringEncodingUTF32;
+/**
+ * The full set of possible string encodings, can be extended by third-party code.
+ */
+typedef StringEncoding = StringEncodingASCII |
+                         StringEncodingUTF8 |
+                         StringEncodingUTF16 |
+                         StringEncodingUTF32 |
+                         StringEncodingCustom;
+
+/**
+ * Character type, stores a unicode code point.
+ */
+typedef Unichar = U32;
+
+// Placeholder until we have syntax for value types.
+value Range
+{
+	U64 index;
+	U64 length;
+}
+
+/**
+ * Interface for string types.  Defines only the primitive operations that all
+ * strings must implement.
+ */
+interface String
+{
+	/**
+	 * Return the length of the string in Unicode code points.
+	 */
+	func length(self: readonly & String) : size_t;
+
+	/**
+	 * The encoding that this string can most easily convert to.  This will
+	 * typically be the encoding that the string uses internally.
+	 * Note that a return value of `StringEncodingASCII` implies
+	 * `StringEncodingUTF8` for all read operations because ASCII is a strict
+	 * superset of UTF-8.
+	 */
+	func fastest_encoding(self: readonly & String) : StringEncoding;
```

Comment:
Are we returning the encoded string, or is the intention to return the type as a value?

Reply:
I was thinking of encodings as being akin to Pony primitives but have almost certainly got the syntax wrong for how we will do this in Verona.

Reply:
That makes sense. We get the instance of the singleton type back. Perfect.

Reply:
Do we yet have syntax for the equivalent of Pony's primitive?

Reply:
Not defined syntax yet, but I'm working with string names for intrinsics, ex. Verona's `"std.addf"(x, y)` ->  MLIR's `%n = std.addf %x, %y`.